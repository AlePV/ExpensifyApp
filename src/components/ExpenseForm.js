// Using this re-usable component for AddExpense and EditExpense


import React from "react";
import moment from "moment";
import {SingleDatePicker} from "react-dates";



// const now = moment();
// console.log(now.format("MMMM D, YYYY"));

export default class ExpenseForm extends React.Component {
    
    constructor(props) {
        super(props);
        
        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? (props.expense.amount / 100).toString() : "",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ""
        };
    }


    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };


    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(()=> ({note}));
    };


    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    };


    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({createdAt}))
        }
    };


    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}))
    };


    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({error: "Please provide a description and an amount"}))
        } else {
            this.setState(() => ({error: ""}))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };


    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>

                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />

                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />

                    <SingleDatePicker 
                        date = {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        focused = {this.state.calendarFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange = {() => false}
                    />

                    <textarea
                        placeholder="Add a note (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>

                    <button>Add Expense</button>

                </form>
            </div>
        )
    }
}


// PLACEHOLDER --> something to show in the input when there is no text typed by the user



// AUTOFOCUS --> makes sure that when we visit the page, automatically puts the cursor and the focus on the input
//          '--> can only do it one input at a time



// Use local component states to track changes --> only when user submits form, we'll do something with the info



// -----------------------------------> STATE NOTES <-----------------------------------
// VALUE={THIS.STATE.DESCRIPTION} --> sets value for this text input equal to the current state value --
//                               '--> but this creates a view-only input 
//                               '--> so we have to write onChange



// -----------------------------------> FUNCTION W/ STATE NOTES <-----------------------------------
// STATE = {DESCRIPTION: ""}  --> first set up default state (for the variable I want, in this case description)
// (e) --> need access to it because it is where the value lives
// CONST DESCRIPTION = E.TARGET.VALUE --> create a const called description = getting its value from what the user types
// THIS.SETSTATE(()=> ({ description })) --> passed inside an updater function and implicitly return an object
//                           '--> here I set {description} equals to the description variable from above
//                           '--> setting description (from my code) equals to what the user types
// -----> REPEATED THIS PROCESS FOR ON-NOTE-CHANGE <-----



// -----------------------------------> REGULAR EXPRESSIONS NOTES <-----------------------------------
// *** changed the type of input from "number" to "text" (line 43) --> so I can restrict the user to write only 2 decimal points
// REGULAR EXPRESSIONS --> A regular expression (regex or regexp for short) is a special text string for describing a search pattern.
// https://regex101.com/r/pqwNbY/1
// IF (AMOUNT.MATCH(/^\d*(\.\d{0,2})?$/)) {THIS.SETSTATE(() => ({AMOUNT}));} --> allows us to ensure that the number comes in a specific format
// We could set this ^^^ formula later, in the submit part --> BUT by doing this here, I ensure the user is not able to type anything else
// (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) --> means that if there is no amount or if it's a match: do what follows
//                                                  '--> Made a change so we don't allow the user to star with a decimal point
//                                                  '--> Made a change so the user can clear the input he wrote (because with the up ^^ change it didn't allow to delete)



// -----------------------------------> MOMENT NOTES <-----------------------------------
// CONST NOW = MOMENT() --> using part of the library "moment" * need to import it before
//                      '--> shows the functions "moment" has
// NOW.FORMAT("MMMM D, YYYY") --> prints the present day in the format of month day, year (Feb 5, 2018)
// CREATEDAT: MOMENT() --> starts at now --> when the user loads the page, that's the day it will show up
//                    '--> that's why createdAt is created with a new instance of moment



// -----------------------------------> DATE PICKER NOTES <-----------------------------------
// <SINGLEDATEPICKER/>--> new instance of SingleDatePicker --> only picks one date (the date at which the event was created)
// DATE --> moment object that represents where you want to start
// ONDATECHANGE --> provide a handler for when the user picks a new date
//             = {THIS.ONDATECHANGE}'--> function that is called when someone picks a new date in the calendar
// FOCUSED --> provide whether or not the input is focused
// ONFOCUSCHANGE --> provide a handler for when the React date's library needs to change that value
// ONDATECHANGE = (CREATEDAT) => { THIS.SETSTATE(() => ({CREATEDAT}))} --> I set the date for createdAt (by implicitly returning {createdAt})
//                    '--> This is the date we want --> mine is in the createdAt variable/object
// --> when the user interacts with the calendar, the third-party library tells you that it's going to close, what we do is provide a handler
// onFocusChange = ({focused}) => { this.setState(() => ({calendarFocused: focused}))}; --> destructured ({focus})
//                                                              '--> set equal the value that we have to the one that it's typed by the user
// NUMBER-OF-MONTHS --> I can determine how many months I want to show in the display
//                 '--> this is actually a function that is included in the redux-dates functions (see docs in safari page to see how it's written)
// IS-OUTSIDE-RANGE = {()=> FALSE-> makes every day available (even the ones before the present day (the day the user enters)
// When we set a SingleDatPicker we have to provide an initial value --> that's why we used MomentJS, to define the current or present day as the starter
//                                                                  '--> get current time in an object moment
//  if (createdAt) {this.setState(() => ({createdAt}))}
//          '--> this means that if there is a date, do the following / if not, don't do anything
//          '--> so the user cannot eliminate the date field completely 
//          '--> prevents the user from being able to clear the date field



// -----------------------------------> ON SUBMIT FUNC. NOTES <-----------------------------------
// ONSUBMIT={THIS.ONSUBMIT} --> this is an onSubmit handler for the form --> used an onSubmit prop on the <form>
//                         '--> on submit we are going to run some code
// ONSUBMIT = (E) => {E.PREVENT-DEFAULT()} --> For our FORMS it's crucial to provide the (e) 
//                             '--> because we need to call e.preventDefault() (which prevents the browser to do the full page refresh)
//                             '--> Allows us to handle it with JavaScript instead
// if (!this.state.description || !this.state.amount) {this.setState(() => ({error: "Please provide a description and an amount"}))} 
//              '-->            '-->                    if there is no description : print error message
//                              '-->          '-->      if there is no amount : print error message
// {this.state.error && <p>{this.state.error}</p>} --> means that if there es an error (if it exists) --> print what the error is
//                                                '--> with the conditional logic from before, if there is an error it will show up

// THIS.PROPS.ONSUBMIT({ DESCRIPTION: {THIS.STATE.DESCRIPTION}}) --> calls the function onSubmit with some data
//                                                              '--> which was created in the AddExpense file and pased on to here)
//                                                              '--> call it with an object with all of these properties
//                                                 '--> Description gets its value directly from the actual description
// AMOUNT: PARSEFLOAT(THIS.STATE.AMOUNT, 10) * 100 --> Takes string number and changes it to a float number 
//                                                '--> Keeps the decimal points, sets the base to ten & eliminates the cents with th *100
// CREATEDAT: THIS.STATE.CREATEDAT.VALUEOF() --> Used to get the time stamp --> because createdAt is in a moment object 
//                                          '--> so to take the moment object out --> use valueOf


// -----------------------------------> CONSTRUCTOR NOTES <-----------------------------------
// Changing from STATE = {} --> to CONSTRUCTOR() = {} --> allows to be used in both EditExpense (which needs the constructor) && AddExpense (which doesn't need it but works the same with it)
// CONSTRUCTOR (PROPS) { SUPER(); THIS.STATE = {}} 
//     '--> CONSTRUCTOR --> Special method for creating and initializing an object created within a class
//                     '--> If no constructor is defined, a default constructor is used
//                '--> Arguments it takes
//                         '--> SUPER --> calls the constructor of a parent class
// -----> DESCRIPTION: PROPS.EXPENSE ? PROPS.EXPENSE.DESCRIPTION : ""
//                                            '--> Check if props.expense exists
//                                                         '--> If it does exist --> start the description value at the same value (as it is, the one the user entered)
//                                                         '--> If it doesn't exist --> start it off as an empty string
// -----> NOTE --> uses the same process ^^^
// -----> AMOUNT --> / 100 because it is a digit with cents (& I don't want the cents)
//              '--> .TO-STRING --> to convert the number to a string (because it's what the component uses)
// -----> CREATED-AT --> Creates an instance of moment at a specific point in time
//                  '--> Need to pass the time stamp in (in my case it is stroed at props.expense.createdAt)