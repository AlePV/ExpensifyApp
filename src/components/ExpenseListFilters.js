import React from "react";
import {connect} from "react-redux";
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from "../actions/filters";
import {DateRangePicker} from "react-dates";

// means this will have access to props.filters.text, and use that as the value of the input
// By changing a stateless function component to a REACT COMPONENT I need to change from {props.filters.text} --> to --> {this.props.filters.text}
export class ExpenseListFilters extends React.Component {
    // Use State to track something in this component
    // The FIVE props of <DateRangePicker/> are required to use it (know which to put because they come in their webpage)
    state = {
        calendarFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    // Going to get called by the React dates library
    // Called with an object that has a start date and an end date
    // '--> Had to de structure startDate and endDate because they are inside the same props.filter object (in the filters file)
    // '--> These two actions were created in the filters file before 
    // '--> Passed the new start date

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    };

    // FOCUSED-INPUT --> just call the state (the thing we are tracking ***check web)
    // ON-FOCUS-CHANGE --> calls new value and sets new value --> justo do it like this, how it says in the webpage ***check web

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate()
        } else if (e.target.value === "amount") {
            this.props.sortByAmount()
        }
    }

    render () {
        return (
            <div className="content-container">

                <div className="input-group">

                <div className="input-group__item">
                    <input 
                        type="text" 
                        placeholder="Search expenses"
                        className="text-input"
                        value={this.props.filters.text} 
                        onChange={this.onTextChange}
                    />
                </div>

                <div className="input-group__item">
                    <select
                        className="select" 
                        value={this.props.filters.sortBy} 
                        onChange={this.onSortChange}>
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                    </select>
                </div>

                <div className="input-group__item">
                    <DateRangePicker
                        startDate = {this.props.filters.startDate}
                        endDate = {this.props.filters.endDate}
                        onDatesChange = {this.onDatesChange}
                        focusedInput = {this.state.calendarFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange = {() => false}
                        showClearDates = {true}
                    />
                </div>

                </div>

            </div>
        );
    }
}

// this is known as a CONTROLLED INPUT -- input where the value is controlled by javascript

// ^^^^^ it's a dropdown for users to pick how they want to sort their expenses (sortBy, from filters) 
// --> <select>   <option></option>   </select>
//  value="date" means the value we'll be using behind the scenes, while "Date" is the value the user sees
// e.target.value is what the user enters or types



// -----------------------------------> UNDERSTANDING (E) NOTES <-----------------------------------
// ONCHANGE HANDLER: every time the input changes = function fires
// (e) => means event argument - which we need to get value
// "target" => means input
// need to use "dispatch" to update the store ==> so keystrokes actually result in a change of input
// in dispatch - pass thing we want to do, the action object
// e.target.value => prints value we want to screen
// WHOLE FUNCTION READS AND WRITES TO STORE // last part dispatchs to change redux store



// -----------------------------------> DATE PICKER NOTES <-----------------------------------
// Allows you to pick the range of dates you want to see

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

// first ()            ^^ takes map state to props func
// second ()                   ^^^^^^^^^^ takes our component
// determine what we want off of the store



// <input type="text"/> ---> this, creates a user input (space where user can enter text )

//make sure input matches value in redux store --> so connect expenseListFilters to the store