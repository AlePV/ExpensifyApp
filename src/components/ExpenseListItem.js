import React from "react";
import {Link} from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

// used de-structuring to grab all the props \\//\\//
const ExpenseListItem = ({id, description, amount, createdAt}) => (
    <div>
    <Link to={`/edit/${id}`} >
        <h3>{description}</h3>
    </Link>
    <p>
        {numeral(amount / 100).format("$0,0.00")} 
            - 
        {moment(createdAt).format("MMMM Do, YYYY")}</p>
    </div>
);

// on click we want to dispatch an action

export default ExpenseListItem;

// <Link to={`/edit/${id}`}>
//             '--> Not a static string because I want to inject the id value **REMEMBER the url of the edit expense page is /edit/id (some number)
//             '--> It is a template string with a javascript expression