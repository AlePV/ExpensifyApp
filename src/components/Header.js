import React from "react";
import {NavLink} from "react-router-dom";
import {connect} from "react-redux";
import {startLogout} from "../actions/auth";

export const Header = ({startLogout}) => (
    <header>
        <h1>Expensify</h1>
            <p> <NavLink to="/dashboard" activeClassName="isActive">Home</NavLink> </p>
            <p> <NavLink to="/create" activeClassName="isActive">Add Expense</NavLink> </p>
            <button onClick={startLogout}>Logout</button>
    </header>
);

// for the bold letters --> shows where you are
const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);

// Create private state for user data --> Private Firebase Data
// '--> Change the firebase rules
// '--> user/userUid/expenses/idOfExpense