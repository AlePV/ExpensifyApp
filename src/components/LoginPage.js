import React from "react";
import {connect} from "react-redux";
import {startLogin} from "../actions/auth";

export const LoginPage = ({startLogin}) => (

    <div>
        <button onClick={startLogin}>Login</button>
    </div>
       
);

const mapDispatchToProps = (dispatch) => ({
// implicitly return an object
    startLogin: () => dispatch(startLogin())
// startLogin prop dispatches the startLogin action
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
// Component doesn't need access to anything state related  --> NO mapStateToProps
// Since I want to dispatch startLogin --> mapDispatchToProps


// If someone is already logged in --> can view private info without needing to relog in every single time