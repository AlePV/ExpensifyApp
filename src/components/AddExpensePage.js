// This adds expenses from the userinterface
// Using the same re-usable component for AddExpense and EditExpense

import React from "react";
import ExpenseForm from "./ExpenseForm";
import {connect} from "react-redux";
import {startAddExpense} from "../actions/expenses";

/* THIS IS THE ORIGINAL COMPONENT
const AddExpensePage = (props) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm
            onSubmit={(expense) => {
                props.dispatch(addExpense(expense));
                props.history.push("/");
            }}
        /> 
    </div>
);

export default connect()(AddExpensePage);
*/

// ONSUBMIT = { (EXPENSE) => {CONSOLE.LOG(EXPENSE);}} --> The function gets called when the form is submitted with valid data
//                                                   '--> We get the expense object back (with all the properties on it - description, amount, note, createdAt)
//                                            '--> We call "expense" --> make sure to be shown in screen 
//                                            '--> This prints what the user typed

// --------------> This file needs to dispatch the given action to the redux store


// IMPORT {CONNECT} FROM "REACT-REDUX" --> This connects the component to the redux store, so it can dispatch
//                                    '--> Using this means that I have access to props.dispatch
// CONNECT()()
//         '--> I don't need anything from the state (first parenthesis)
//           '--> Pass component (second parenthesis)
// --> Try to dispatch an action
// IMPORT {ADDEXPENSE} --> It is an action generator that I already created
// PROPS.DISPATCH(ADDEXPENSE(EXPENSE)) --> Takes (expense) argument and pass it through

// --> Automate redirect (it doesn't do it without this function)

// PROPS.HISTORY.PUSH("/") --> Func to programarily change pages (automatic re-direct)
//                        '--> Does not go through entire page refresh, switches over using browser routing
//                     '--> Takes only one string: the path



// Ways to return your dispatcher functions, allowing you to abstract them away from the component itself

// NEW COMPONENT TO MAKE IT EASIER TO TEST

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push("/");
    };
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);