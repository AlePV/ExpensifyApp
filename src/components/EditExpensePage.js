// Using the same re-usable component for AddExpense and EditExpense


import React from "react";
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {startEditExpense, startRemoveExpense} from "../actions/expenses"; 

// ORIGINAL FILE -----> BEFORE CHANGING IT TO TEST IT EASIER
/*
const EditExpensePage = (props) => {
    console.log(props);
    return (
        <div>
            <ExpenseForm 
                expense = {props.expense}
                onSubmit = {(expense) => {
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push("/");
                }}
            />
            <button onClick={() => {
                props.dispatch(removeExpense({id: props.expense.id}));
                props.history.push("/");
            } }>Remove</button>
        </div>
    );
};


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};


export default connect(mapStateToProps)(EditExpensePage);
*/

// DIFFERENCE BETWEEN ADDEXPENSE & EDITEXPENSE --> AddExpense uses empty defaults
//                                            '--> EditExpense needs current values for an expense
//                                                 '--> So,to change this I need a single prop that gets passed to ExpenseForm
//                                                 '--> Conditional because if it doesn't exist, not entered by the user, I will stay with the default



// -----------------------------------> MAP-STATE-TO-PROPS NOTES <-----------------------------------
// CONST MAP-STATE-TO-PROPS = (STATE, PROPS) => { RETURN {EXPENSE} } 
//                                       '--> I want to be able to give the component the current expense object --> because it will EDIT it
//                               '--> STATE is where the expenses array lives
//                               '--> I'm searching for an expenses array whose an id matches this props.match.params.id
//                                    '--> I have access to it with the (props)
//                                    '--> Now, id is equal to the URL ID
//                               '--> Added one new prop: expense --> uses current props to search the expenses array
//                               '--> FIND ((EXPENSE) => {RETURN TRUE})
//                                        '--> allows us to search through an array looking for a single item
//                                        '--> We determine whether or not we found the correct item by returning TRUE from the callback
//                                        '--> Calls a function --> gets the individual item (expense)
//                                        '--> RETURN TRUE --> the first item it searches through will be considered a match (& that expense would get passed in here)
//                                                        '--> IF I return FALSE it would never find a match
//                                        EXPENSE.ID === PROPS.MATCH.PARAMS.ID --> Checks if the expense's id are equal, if they are, then we have a match 
//                                                                            '--> want to take that object and put it on the new prop EXPENSE



// React Router renders our Higher Order Component (HOC) --> The HOC passes the props through and allows us to add new ones

// Render ExpenseForm in the EditExpensePage --> ExpenseForm takes in a prop : OnSubmit
//                                          '--> Populate the fields with their existing values --> To do this: take expense and pass it down
//                                          '--> Define a new prop, an existing expense that we have to access to with {props.expense}

// EditExpensePage
// Editing the expense with the id of {props.match.params.id}
// Want to only use defaults if no expense was passed down
// If an expense is passed down, start & use those values



// -----------------------------------> PROPS.DISPATCH(EDIT-EXPENSE()) NOTES <-----------------------------------
// PROPS.DISPATCH(EDIT-EXPENSE(PROPS.EXPENSE.ID, EXPENSE))
//          '--> To call an already created action, need to starts with props.dispatch
//                     '--> Edit expense action takes two parameters: id and updates
//                                     '--> ID --> is stored in props.expense.id
//                                     '--> EXPENSE --> the updates are stored in the first argument (from above) 
//                                                 '--> because for the EditExpense page I stored the updates in a new prop called expense
//                                                 '--> It was stored in EXPENSE = {PROPS.EXPENSE}

// The param tries to get an id from an id variable (but we don't have it)
// '--> I changed this to ID: PROPS.EXPENSE.ID (which we have)


export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    };

    onRemove = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.props.history.push("/");
    };
    render() {
        return (
            <div>
            <ExpenseForm 
            expense = {this.props.expense}
            onSubmit = {this.onSubmit}
            />
            <button 
                onClick={this.onRemove}>Remove</button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        startEditExpense: (expense, id) => dispatch(startEditExpense(expense, id)),
        startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);