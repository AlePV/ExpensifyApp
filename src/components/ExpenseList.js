import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

// UNCONNECTED
// <h1>ExpenseList</h1> --> I DELETED THIS
export const ExpenseList = (props) => (
    <div> 
    {
        props.expenses.length === 0 ? (
            <p>No expenses</p>
        ) : (
            props.expenses.map ((expense) => {
                return <ExpenseListItem key={expense.id} {...expense}/>
            })
        )
    }
    </div>
    //                            ^^^ always need to put key by default
    //                                               ^^^ I spread expense, so I can have access to any of the characteristics it has inside
    //              ^^^^ .map returns to screen each individual item of an array :: needs an instance of it, which is store in "ExpenseListItem"
    // {props.filters.text}
);

// info that we want our component to access from the store -- & first the store state

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
        // Instead of writing the below info, write this ^^^^ which passes the visibleExpenses 
        // func (so I only see the ones I have stated before), which is sorted and filtered
        // expenses: state.expenses,
        // filters: state.filters
    };
};

export default connect(mapStateToProps)(ExpenseList);

// THIS SYNTAX IS NOT THAT COMMON
/*
const ConnectedExpenseList = connect((state) => {
    return {
        expenses: state.expenses
    };
})(ExpenseList);

export default ConnectedExpenseList;
*/ 


// maps store state to component props