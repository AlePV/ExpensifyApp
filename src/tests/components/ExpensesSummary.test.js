import React from "react";
import {shallow} from "enzyme";
import {ExpensesSummary} from "../../components/ExpensesSummary";

test("should render correctly ExpensesSummary with 1 expense", () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={500}/>);
    expect(wrapper).toMatchSnapshot();
});

test("should render correctly ExpensesSummary with 2 expenses", () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={2} expensesTotal={5000}/>);
    expect(wrapper).toMatchSnapshot();
});