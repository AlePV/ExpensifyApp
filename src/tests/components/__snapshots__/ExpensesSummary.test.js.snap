// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly ExpensesSummary with 1 expense 1`] = `
<div>
  <h3>
    Viewing 
    1
     
    expense
    , total: 
    $5.00
  </h3>
</div>
`;

exports[`should render correctly ExpensesSummary with 2 expenses 1`] = `
<div>
  <h3>
    Viewing 
    2
     
    expenses
    , total: 
    $50.00
  </h3>
</div>
`;
